k2/csrc/ragged.cu

```diff
@@ -440,32 +451,105 @@ RaggedShape ComposeRaggedShapes(const RaggedShape &a, const RaggedShape &b) {
+  K2_CHECK(ctx1->IsCompatible(*ctx2));
+
+  // check row_splits and row_ids of axis-1
+  if (cached_tot_size1 != -1) {
+    if (row_ids1 != nullptr) K2_CHECK_EQ(cached_tot_size1, row_ids1->Dim());
+    if (row_splits1 != nullptr) {
+      // may be slow as it may copy memory from device to host
+      K2_CHECK_EQ(cached_tot_size1, row_splits1->Back());
+    }
+  }
+
+  // check row_splits and row_ids of axis-2
+  if (cached_tot_size2 != -1) {
+    if (row_ids2 != nullptr) K2_CHECK_EQ(cached_tot_size2, row_ids2->Dim());
+    if (row_splits2 != nullptr) {
+      // may be slow as it may copy memory from device to host
+      K2_CHECK_EQ(cached_tot_size2, row_splits2->Back());
+    }
+  }
+
+  std::vector<RaggedShapeDim> axes(2);
+  // set row_splits and row_ids for axis 1
+  if (row_splits1 != nullptr) {
+    axes[0].row_splits = *row_splits1;
+  } else {
+    // work out row_splits1, see code in RaggedShape2 above for the reason
+    int32_t num_rows = row_ids1->Dim() == 0 ? 0 : row_ids1->Back() + 1;
+    Array1<int32_t> row_splits_array(ctx1, num_rows + 1);
+    RowIdsToRowSplits(*row_ids1, row_splits_array);
+    axes[0].row_splits = row_splits_array;
+  }
+  if (row_ids1 != nullptr) axes[0].row_ids = *row_ids1;
+  axes[0].cached_tot_size = cached_tot_size1;
+
+  // set row_splits and row_ids for axis 2
+  if (row_splits2 != nullptr) {
+    axes[1].row_splits = *row_splits1;
+  } else {
+    // work out row_splits1, see code in RaggedShape2 above for the reason
+    int32_t num_rows = row_ids2->Dim() == 0 ? 0 : row_ids2->Back() + 1;
+    Array1<int32_t> row_splits_array(ctx1, num_rows + 1);
+    RowIdsToRowSplits(*row_ids2, row_splits_array);
+    axes[1].row_splits = row_splits_array;
+  }
+  if (row_ids2 != nullptr) axes[1].row_ids = *row_ids2;
+  axes[1].cached_tot_size = cached_tot_size2;
+
+  // we don't check here if
+  // row_splits1[row_splits1.Dim() - 1] == row_ids1.Dim()
+  //   == (row_splits2.Dim() - 1)
+  //   >= (row_ids2[row_ids2.Dim() - 1] + 1)
+  // but RaggedShape(axes) below will check this.
+  return RaggedShape(axes);
+}
+
+RaggedShape RaggedShapeFromTotSizes(ContextPtr &c, int32_t num_axes,
+                                    int32_t *tot_sizes) {
+  K2_CHECK_GE(num_axes, 2);
+  std::vector<RaggedShapeDim> axes(num_axes - 1);
+  // In future we might choose to allocate everything in one big array, to avoid
+  // multiple allocations, but for now just do it the simple way.
+  for (int32_t axis = 1; axis < num_axes; axis++) {
+    axes[axis - 1].row_splits = Array1<int32_t>(c, tot_sizes[axis - 1] + 1);
+    axes[axis - 1].row_ids = Array1<int32_t>(c, tot_sizes[axis]);
+    axes[axis - 1].cached_tot_size = tot_sizes[axis];
+  }
+  return RaggedShape(axes);
 }
 
 Array1<int32_t *> GetRowSplitsPtr(RaggedShape &src) {
-  Array1<int32_t *> array;
-  // TODO(haowen): implement
-  return array;
+  int32_t axes = src.NumAxes();
+  std::vector<int32_t *> row_splits_start(axes - 1);
```

Comment:
Is there any constraint for the value of `src.NumAxes()`?

Reply:
Sure, added >=2 check